src/syncerd/bitcoin_syncer.rs

```diff
@@ -411,6 +413,107 @@ fn query_addr_history(
+        Some(address_type) => {
+            return Err(Error::Farcaster(format!(
+                "Sweeping addresses only supports native segwit v1 addresses. Address has type: {}",
+                address_type
+            )));
+        }
+        None => return Err(Error::Farcaster(format!("Invalid to be swept address"))),
+    }
+
+    let sk = bitcoin::PrivateKey::from_slice(&private_key, bitcoin::Network::Testnet)?;
+    let pk = bitcoin::PublicKey::from_private_key(bitcoin::secp256k1::SECP256K1, &sk);
+
+    let client = Client::new(&electrum_server)?;
+    let unspent_txs = client.script_list_unspent(&address.script_pubkey())?;
+
+    if unspent_txs.len() == 0 {
+        return Ok(vec![]);
+    }
+
+    let in_amount = unspent_txs
+        .iter()
+        .fold(0, |acc, unspent_tx| acc + unspent_tx.value);
+    let inputs: Vec<bitcoin::TxIn> = unspent_txs
+        .iter()
+        .map(|unspent_output| bitcoin::TxIn {
+            previous_output: bitcoin::OutPoint {
+                txid: unspent_output.tx_hash,
+                vout: unspent_output.tx_pos as u32,
+            },
+            script_sig: bitcoin::Script::default(),
+            sequence: (1 << 31) as u32,
+            witness: bitcoin::Witness::new(),
+        })
+        .collect();
+
+    let mut unsigned_tx = bitcoin::Transaction {
+        version: 2,
+        lock_time: 0,
+        input: inputs,
+        output: vec![bitcoin::TxOut {
+            value: in_amount,
+            script_pubkey: destination_address.script_pubkey(),
+        }],
+    };
+    unsigned_tx.output[0].value = in_amount - 2 * unsigned_tx.vsize() as u64;
+    let mut psbt = bitcoin::util::psbt::PartiallySignedTransaction::from_unsigned_tx(unsigned_tx)
+        .map_err(|_| Error::Syncer(SyncerError::InvalidPsbt))?;
+    psbt.outputs[0].witness_script = Some(destination_address.script_pubkey());
+
+    // sign the inputs and collect the witness data
+    for (index, input) in psbt.inputs.iter_mut().enumerate() {
+        input.witness_utxo = Some(bitcoin::TxOut {
+            value: unspent_txs[index].value,
+            script_pubkey: address.script_pubkey(),
+        });
+        let script = p2wpkh_script_code(&address.script_pubkey());
+        input.witness_script = Some(script.clone());
+        let txin = TxInRef::new(&psbt.unsigned_tx, index);
+        let sig_hash = signature_hash(
+            txin,
+            &script,
+            unspent_txs[index].value,
+            bitcoin::EcdsaSighashType::All,
+        );
+        let message = bitcoin::secp256k1::Message::from_slice(&sig_hash)?;
+        let signature = bitcoin::secp256k1::SECP256K1.sign_ecdsa(&message, &sk.inner);
+        let sig_all = bitcoin::util::ecdsa::EcdsaSig::sighash_all(signature);
+        input.partial_sigs.insert(pk, sig_all);
+        let (pubkey, full_sig) = input.partial_sigs.iter().next().ok_or(Error::Farcaster(
+            "to be signed sweep input empty".to_string(),
+        ))?;
```

Comment:
this error is unreachable, insert is just above, and `pubkey`, `full_sig` are just `pk`, `sig_all`

Reply:
Oh, this was ugly. Done in 056d1843ba99762556f6cd540faa379a3f911ee8